src/pgas.jl

```diff
@@ -0,0 +1,142 @@
+""" 
+    initialization(model::AbstractStateSpaceModel)
+
+Define the distribution of the initial state of the State Space Model
+"""
+function initialization end
+
+""" 
+    transition(model::AbstractStateSpaceModel, state, step)
+
+Define the transition density of the State Space Model
+Must return `nothing` if it consumed all the data
+"""
+function transition end
+
+"""
+    observation(model::AbstractStateSpaceModel, state, step)
+
+Return the log-likelihood of the observed measurement conditional on the current state of the model. 
+Must return `nothing` if it consumed all the data
+"""
+function observation end
+
+""" 
+    isdone(model::AbstractStateSpaceModel, step)
+
+Return `true` if model reached final state else `false`
+"""
+function isdone end
+
+"""
+    previous_state(trace::SSMTrace)
+
+Return `Xₜ₋₁` or `nothing` from `model`
+"""
+function previous_state(trace::SSMTrace)
+    return trace.model.X[step(trace) - 1]
+end
+
+function past_idx(trace::SSMTrace)
+    return 1:(step(trace) - 1)
+end
+
+"""
+    step(model::AbstractStateSpaceModel)
```

Comment:
I guess `step` could be confusing since then one has two different functions `AbstractMCMC.step` and `AdvancedPS.step`.